iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndVectorizeLinalgTensorOps2.cpp

```diff
@@ -0,0 +1,333 @@
+#include "mlir/Dialect/Linalg/Transforms/Hoisting.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/MemRef/Transforms/Passes.h"
+#include "mlir/Dialect/SCF/Transforms.h"
+#include "mlir/Dialect/Vector/VectorTransforms.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-llvmcpu-tile-and-vectorize"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+// Could just be linalg::TilingPattern with a ContractionOpInterface filter, but
+// that is always templated on an op.
+struct TileWorkgroups : public linalg::LinalgBaseTilingPattern {
+  using Base = linalg::LinalgBaseTilingPattern;
+  TileWorkgroups(MLIRContext *context, linalg::LinalgTilingOptions options,
+                 linalg::LinalgTransformationFilter marker)
+      : LinalgBaseTilingPattern(context, options, marker) {}
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    auto contractionOp = dyn_cast<linalg::ContractionOpInterface>(op);
+    if (!contractionOp) return failure();
+
+    linalg::TiledLinalgOp tiledLinalgOp;
+    if (failed(Base::matchAndRewriteBase(op, rewriter, tiledLinalgOp))) {
+      return failure();
+    }
+    rewriter.replaceOp(op, tiledLinalgOp.tensorResults);
+    return success();
+  }
+};
+
+}  // namespace
+
+namespace {
+struct LLVMCPUTileAndVectorizePass2
+    : public LLVMCPUTileAndVectorize2Base<LLVMCPUTileAndVectorizePass2> {
+  LLVMCPUTileAndVectorizePass2(bool vectorize = true)
+      : lowerToVectors(vectorize) {}
+  LLVMCPUTileAndVectorizePass2(const LLVMCPUTileAndVectorizePass2 &pass) {
+    lowerToVectors = pass.lowerToVectors;
+  }
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect, memref::MemRefDialect,
+                    vector::VectorDialect>();
+  }
+  void runOnOperation() override;
+
+ private:
+  bool lowerToVectors;
+};
+}  // namespace
+
+void LLVMCPUTileAndVectorizePass2::runOnOperation() {
+  MLIRContext *context = &getContext();
+  auto funcOp = getOperation();
+
+  DEBUG_WITH_TYPE(DEBUG_TYPE, {
+    llvm::dbgs() << "\n--- Before LLVMCPUTileAndVectorizePass2 ---\n";
+    funcOp.print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
+    llvm::dbgs() << "\n\n";
+  });
+
+  // Assume there is a single op with a lowering config we use to drive the
+  // tiling decisions.
+  IREE::Codegen::LoweringConfigAttr config;
+  funcOp.walk([&](linalg::LinalgOp linalgOp) {
+    if (auto opConfig = getLoweringConfig(linalgOp)) {
+      if (opConfig) {
+        // Duplicate configurations.
+        if (config) return signalPassFailure();
+        config = opConfig;
+      }
+    }
+  });
+
+  // Skip vetorization if not all the ops can be vectorized.
+  bool isVectorizable = true;
+  {
+    SmallVector<Operation *> computeOps;
```

Comment:
I would drop this right now, or just check this and fail. Ideally we want to catch this decision early in `KernelConfig.cpp`, so this logic should move there, so that we are gauranteed to vectorize if we go down this path.

Reply:
I added a TODO here, and plan to move it to `KernelConfig.cpp` soon.

Instead of signaling failure, I think it's fine to do nothing. Because it is under development, and I don't expect any others to use it. It allows us to compile most of kernels successfully. I can signal the failure if you strongly think we should do that.